File: packages/youtube-player/ios/Plugin/Classes/Options/CreatePlayerOptions.swift
import Capacitor
import Foundation

@objc public class CreatePlayerOptions: NSObject {
    let autoplay: Bool
    let ccLoadPolicy: Bool
    let controls: Bool
    let end: Int?
    let frame: PlayerFrame
    let fullscreen: Bool
    let id: String
    let ivLoadPolicy: Bool
    let mute: Bool
    let rel: Bool
    let start: Int?
    let videoId: String?

    init(_ call: CAPPluginCall) throws {
        let playerOptions = call.getObject("options")
        self.autoplay = playerOptions?["autoplay"] as? Bool ?? false
        self.ccLoadPolicy = playerOptions?["ccLoadPolicy"] as? Bool ?? false
        self.controls = playerOptions?["controls"] as? Bool ?? true
        self.end = CreatePlayerOptions.getSecondsFromObject(playerOptions, "end")
        self.frame = try PlayerFrame.getFrameFromCall(call)
        self.fullscreen = playerOptions?["fullscreen"] as? Bool ?? false
        self.id = call.getString("id") ?? UUID().uuidString
        self.ivLoadPolicy = playerOptions?["ivLoadPolicy"] as? Bool ?? false
        self.mute = playerOptions?["mute"] as? Bool ?? false
        self.rel = playerOptions?["rel"] as? Bool ?? false
        self.start = CreatePlayerOptions.getSecondsFromObject(playerOptions, "start")
        self.videoId = call.getString("videoId")
    }

    private static func getSecondsFromObject(_ object: JSObject?, _ key: String) -> Int? {
        if let value = object?[key] as? Int {
            return value
        }
        if let value = object?[key] as? Double {
            return Int(value)
        }
        return nil
    }
}

File: packages/youtube-player/android/src/main/java/io/capawesome/capacitorjs/plugins/youtubeplayer/classes/options/CreatePlayerOptions.java
package io.capawesome.capacitorjs.plugins.youtubeplayer.classes.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import io.capawesome.capacitorjs.plugins.youtubeplayer.classes.PlayerFrame;
import java.util.UUID;

public class CreatePlayerOptions {

    private final boolean autoplay;
    private final boolean ccLoadPolicy;
    private final boolean controls;

    @Nullable
    private final Integer end;

    @NonNull
    private final PlayerFrame frame;

    private final boolean fullscreen;

    @NonNull
    private final String id;

    private final boolean ivLoadPolicy;
    private final boolean mute;
    private final boolean rel;

    @Nullable
    private final Integer start;

    @Nullable
    private final String videoId;

    public CreatePlayerOptions(@NonNull PluginCall call) throws Exception {
        JSObject playerOptions = call.getObject("options", null);
        this.autoplay = playerOptions != null && playerOptions.optBoolean("autoplay", false);
        this.ccLoadPolicy = playerOptions != null && playerOptions.optBoolean("ccLoadPolicy", false);
        this.controls = playerOptions == null || playerOptions.optBoolean("controls", true);
        this.end = CreatePlayerOptions.getSecondsFromObject(playerOptions, "end");
        this.frame = PlayerFrame.getFrameFromCall(call);
        this.fullscreen = playerOptions != null && playerOptions.optBoolean("fullscreen", false);
        this.id = CreatePlayerOptions.getIdFromCall(call);
        this.ivLoadPolicy = playerOptions != null && playerOptions.optBoolean("ivLoadPolicy", false);
        this.mute = playerOptions != null && playerOptions.optBoolean("mute", false);
        this.rel = playerOptions != null && playerOptions.optBoolean("rel", false);
        this.start = CreatePlayerOptions.getSecondsFromObject(playerOptions, "start");
        this.videoId = call.getString("videoId");
    }

    public boolean getAutoplay() {
        return autoplay;
    }

    public boolean getCcLoadPolicy() {
        return ccLoadPolicy;
    }

    public boolean getControls() {
        return controls;
    }

    @Nullable
    public Integer getEnd() {
        return end;
    }

    @NonNull
    public PlayerFrame getFrame() {
        return frame;
    }

    public boolean getFullscreen() {
        return fullscreen;
    }

    @NonNull
    public String getId() {
        return id;
    }

    public boolean getIvLoadPolicy() {
        return ivLoadPolicy;
    }

    public boolean getMute() {
        return mute;
    }

    public boolean getRel() {
        return rel;
    }

    @Nullable
    public Integer getStart() {
        return start;
    }

    @Nullable
    public String getVideoId() {
        return videoId;
    }

    @NonNull
    private static String getIdFromCall(@NonNull PluginCall call) {
        String id = call.getString("id");
        if (id == null) {
            id = UUID.randomUUID().toString();
        }
        return id;
    }

    @Nullable
    private static Integer getSecondsFromObject(@Nullable JSObject object, @NonNull String key) {
        if (object == null || !object.has(key)) {
            return null;
        }
        return object.optInt(key);
    }
}
